File: problems/2025/09-07/merge-k-sorted-linked-lists/solution.js
function Node(d) {
  this.data = d;
  this.next = null;
}

function mergeKLists(arr) {
  // Min-heap for nodes by node.data
  const heap = [];
  const less = (a, b) => a.data < b.data;
  const swap = (i, j) => { const t = heap[i]; heap[i] = heap[j]; heap[j] = t; };
  const up = (i) => { while (i > 0) { const p = (i - 1) >> 1; if (!less(heap[i], heap[p])) break; swap(i, p); i = p; } };
  const down = (i) => { const n = heap.length; while (true) { let l = i * 2 + 1, r = l + 1, m = i; if (l < n && less(heap[l], heap[m])) m = l; if (r < n && less(heap[r], heap[m])) m = r; if (m === i) break; swap(i, m); i = m; } };
  const push = (x) => { heap.push(x); up(heap.length - 1); };
  const pop = () => { const last = heap.pop(); if (heap.length === 0) return last; const top = heap[0]; heap[0] = last; down(0); return top; };

  for (const head of arr) if (head) push(head);

  const dummy = new Node(0);
  let tail = dummy;

  while (heap.length > 0) {
    const node = pop();
    tail.next = node;
    tail = node;
    if (node.next) push(node.next);
  }

  tail.next = null;
  return dummy.next;
}

// Helpers
function buildList(a) {
  let head = null, tail = null;
  for (const v of a) {
    const n = new Node(v);
    if (!head) head = tail = n; else { tail.next = n; tail = n; }
  }
  return head;
}

function toArray(head) {
  const out = [];
  while (head) { out.push(head.data); head = head.next; }
  return out;
}

// quick checks
let lists = [buildList([1,3,7]), buildList([2,4,8]), buildList([9])];
let h = mergeKLists(lists);
console.log(toArray(h)); // [1,2,3,4,7,8,9]

lists = [buildList([1,3]), buildList([8]), buildList([4,5,6])];
h = mergeKLists(lists);
console.log(toArray(h)); // [1,3,4,5,6,8]

File: problems/2025/09-07/merge-k-sorted-linked-lists/solution.py
from typing import Optional, List, Tuple
import heapq


class Node:
    def __init__(self, d: int):
        self.data = d
        self.next: 'Optional[Node]' = None


class Solution:
    def mergeKLists(self, arr: List[Optional[Node]]) -> Optional[Node]:
        # Min-heap of tuples: (value, unique_id, node)
        heap: List[Tuple[int, int, Node]] = []
        push_id = 0
        for head in arr:
            if head is not None:
                heapq.heappush(heap, (head.data, push_id, head))
                push_id += 1

        dummy = Node(0)
        tail = dummy

        while heap:
            val, _, node = heapq.heappop(heap)
            tail.next = node
            tail = node
            if node.next is not None:
                heapq.heappush(heap, (node.next.data, push_id, node.next))
                push_id += 1

        # ensure end terminates
        tail.next = None
        return dummy.next


# Helpers for quick checks

def build_list(values: List[int]) -> Optional[Node]:
    head: Optional[Node] = None
    tail: Optional[Node] = None
    for v in values:
        n = Node(v)
        if head is None:
            head = tail = n
        else:
            assert tail is not None
            tail.next = n
            tail = n
    return head


def to_array(head: Optional[Node]) -> List[int]:
    out: List[int] = []
    while head is not None:
        out.append(head.data)
        head = head.next
    return out


if __name__ == "__main__":
    s = Solution()

    lists = [build_list([1, 3, 7]), build_list([2, 4, 8]), build_list([9])]
    h = s.mergeKLists(lists)
    print(to_array(h))  # [1,2,3,4,7,8,9]

    lists = [build_list([1, 3]), build_list([8]), build_list([4, 5, 6])]
    h = s.mergeKLists(lists)
    print(to_array(h))  # [1,3,4,5,6,8]
